timemachine/fe/single_topology.py

```diff
@@ -0,0 +1,866 @@
+            elif dst == atom and src in core_idxs and src != anchor_idx:
+                nbs_2.add((atom, src))
+
+    return nbs_1, nbs_2
+
+
+def setup_dummy_interactions(ff, mol_a, mol_b, core, dummy_group, anchor):
+    """
+    Add restraints between dummy atoms in a dummy_group and core atoms. Interactions involving
+    mol_b's dummy atoms are added in a factorizable, and numerically stable way relative to both
+    mol_b and mol_a.
+
+    Parameters
+    ----------
+
+    mol_a: Chem.Mol
+        mol_a's parameters are mainly used to check for numerical stability.
+
+    mol_b: Chem.Mol
+        Source of the dummy atoms
+
+    core: np.array(N,2)
+        Core atoms
+
+    dummy_group: set
+        A set of dummy atoms tethered to an anchor
+
+    Returns
+    -------
+    (idxs, params)
+        A set of bond, angle, proper, improper, c_angle, and x_angle restraints
+
+    """
+    core_b_to_a = {b: a for a, b in core}
+
+    assert anchor in core_b_to_a, "anchor must be in core"
+
+    dummy_group = list(dummy_group)
+    for d in dummy_group:
+        assert d not in core_b_to_a, "dummy " + str(d) + " must not be in core"
+
+    # these idxs/params can and should be cached, but is repeated here to keep the api simple
+    mol_a_top = topology.BaseTopology(mol_a, ff)
+    mol_b_top = topology.BaseTopology(mol_b, ff)
+
+    mol_a_bond_params, mol_a_hb = mol_a_top.parameterize_harmonic_bond(ff.hb_handle.params)
+    mol_a_angle_params, mol_a_ha = mol_a_top.parameterize_harmonic_angle(ff.ha_handle.params)
+
+    mol_b_bond_params, mol_b_hb = mol_b_top.parameterize_harmonic_bond(ff.hb_handle.params)
+    mol_b_angle_params, mol_b_ha = mol_b_top.parameterize_harmonic_angle(ff.ha_handle.params)
+    mol_b_proper_params, mol_b_pt = mol_b_top.parameterize_proper_torsion(ff.pt_handle.params)
+    mol_b_improper_params, mol_b_it = mol_b_top.parameterize_improper_torsion(ff.it_handle.params)
+
+    mol_a_bond_idxs = mol_a_hb.get_idxs()
+    mol_a_angle_idxs = mol_a_ha.get_idxs()
+
+    mol_b_bond_idxs = mol_b_hb.get_idxs()
+    mol_b_angle_idxs = mol_b_ha.get_idxs()
+    mol_b_proper_idxs = mol_b_pt.get_idxs()
+    mol_b_improper_idxs = mol_b_it.get_idxs()
+
+    mol_b_core = core_b_to_a.keys()
+    mol_b_full_geometry = geometry.classify_geometry(mol_b)
+    mol_b_core_geometry = geometry.classify_geometry(mol_b, core=mol_b_core)
+
+    # pick an arbitrary atom in the dummy_group and find the anchors, there may be
+    # multiple anchors, eg (d=dummy, c=core):
+    #   d...d
+    #   |   |
+    #   c---c
+    root_anchors = dummy.identify_root_anchors(mol_b_bond_idxs, mol_b_core, dummy_group[0])
+    assert anchor in root_anchors
+    # (ytz): we can relax this assertion later on.
+    assert len(root_anchors) == 1, "multiple root anchors found."
+
+    mol_b_ring_bonds = []
+    for b in mol_b.GetBonds():
+        if b.IsInRing():
+            mol_b_ring_bonds.append((b.GetBeginAtomIdx(), b.GetEndAtomIdx()))
```

Comment:
could this reuse `get_romol_bonds`? I.e. something like
```suggestion
    mol_b_ring_bonds = [bond for bond in utils.get_romol_bonds(mol_b) if bond.IsInRing()]
```